### DISCUSSION TITLE
Function-calling React Component

### DISCUSSION THREAD

**Discussion by thorwhalen** [i2mint/oui#15] [Ideas]
## Objective

A reusable react component that essentially calls a remote python function. 

This component could have all kinds of parameters to control different (view and control) aspects of the component, but the only two required parameters are:

* An API that specifies the necessary information to get user inputs to call the function with (related to the signature of the function
* An API to call the functions, given the inputs

Though this itself could be parametrized, we'll say that the "user input" view will be for now a form. 

Therefore, we look into react frameworks that will help us:

* take a declarative approach to form bulding
* reduce boilerplate in creating forms, along with all the desirable functionality (input element flexibility, validation, value binding, etc.)

## Proposed Steps 

1.	Setup a web service that exposes function metadata: You will need to create a web service (say, using Flask, FastAPI or Django) in Python that can provide the signature information for your Python function. The endpoint might return the information in a structured format like JSON. For instance, it could return a list of arguments, their types, and any default values. You might also include other details you want to reflect in your form, such as help text.
2.	Convert function metadata to JSON schema: On your frontend application, fetch the function metadata from your web service. Once you have this information, you can write a function to convert this metadata into a JSON Schema. This will be used to generate your form using react-jsonschema-form. The details of this will depend on exactly what information you have about each argument, but at a minimum, you'd probably map argument names to field names, and argument types to JSON Schema types.
3.	Create a React component using react-jsonschema-form: Now you can use react-jsonschema-form to create a form from your JSON Schema. It will generate a form with fields corresponding to your function arguments, and it will automatically validate user input based on the schema.
4.	Submit the form: When the form is submitted, react-jsonschema-form can give you a JSON object of the form data. You then need to send this data to another endpoint of your web service that will call the Python function with the provided arguments.
5.	Process the result: Once the web service has computed the result of the function, it can return it in the HTTP response. You can then display this result in your frontend application.

## Useful links

* [Example of auto func-to-rjsf](https://github.com/i2mint/ju/blob/029469d64bdd6ab6f23f3f98982f26e1f71a46ae/ju/rjsf.py#L8)
* [Fiddle for react-jsonschema-form](https://rjsf-team.github.io/react-jsonschema-form/)
* [pydantic's json schema conversion](https://docs.pydantic.dev/latest/concepts/json_schema/)
* [Other parts of the original post](https://github.com/otosense/oui/discussions/15#discussioncomment-7253659)

**Comment by thorwhalen** [comment]
## Back-end support to the component

Now let's look zoom out to include the back-end support for such a component. 
First, we'll look at the steps involved in providing the front-end with the necessary information to create the function-calling form. 

Note that the diagrams shown below are abstract data/functional relationships. 
There's no intended detail on implementation, namely on aspects like:
* dynamic/static: is it computed dynamically, or retrieved from a store
* frontend/backend: does "it" (computation or storage) happen on the frontend or the backend?

We give the implementation-free view precisely to help the design be free and separate from those concerns. 

### Simple pipeline

A `func` is selected via a `func_id`. The `func_info` is extracted from this `func` and then used to produce the `json_schema` for this `func`, which is then used (by `react-jsonschema-form`) to produce the "ready to render" ui information.

Implementation notes:
* dynamic/static design choice: The `func_` functionality is probably implemented by retrieving the function from a persisted store (e.g. containing the pickle/dill of the function). Though, one could note that dynamic implementations are also possible, for instance, resolving a dot-path to an object such as `os.path.join` to it's function. 
* The `func_info_` functionality could be implemented dynamically, on the fly, on retrieval of the function (using the `obj_of_data` argument of a [dol.wrap_kvs](https://i2mint.github.io/dol/module_docs/dol/trans.html#dol.trans.wrap_kvs) wrapper, for example). A static implementation would have a store that contains the `func_info` of the store, which could itself come from a dynamic sync with the function store (example, on write (of the function store) or read (of the func_info store)). 
* The `json_schema_` functionality has the same options as the `func_info_`, except here, there's an additional aspect that one may want to consider: The possibility that the relationship between `func_info` and `json_schema` may not be 1-to-1, but 1-to-many. Perhaps the (dev) user might want to render the func_info in several ways (e.g. use a different element to acquire a same argument, or different labels, etc.). This is what we address explicitly in the next section, through the `schema_config` argument.
* The `form_` functionality has the same options as `json_schema_`, including the 1-to-many comment, that leads to the possible need of a `ui_config`, shown in the next section.

We discussed the some of the dynamic/static choices in the implementation notes, but not the frontend/backend choices. 
The only one I see as a valid design choice here is for the `json_schema`. It could be computed, or stored, in the backend, but it could also be computed in the frontend, or stored in a different backend. In both cases, the frontend would get the `func_info` from the first backend, then call a `json_schema_` function -- which point to a JS function to compute the needed data, or use an API to ask a server to do so. 

<img width="139" alt="image" src="https://github.com/otosense/oui/assets/1906276/d86aa4a6-34f1-4c56-8594-e8a21c9e7b24">


### Adding configs

In this diagram, we add optional `schema_config` and a `ui_config` arguments to indicate that perhaps there's (respectively) more than one   `json_schema` for a `func_info` and more than one `form` for a same `json_schema`, and we'd like to leave the choice open for change. Emphasis on "optional". Note that `schema_config` and `ui_config` have a `=` sign in front of them, which indicates that they have a default, therefore the functions they feed into have a default behavior, which is the one that our linear pipeline above would have.

Note that again, there's no indication if, say, the `json_schema_` is dynamically computing the `json_schema` from the `func_info` and `schema_config`, or if it is retrieving it from a (statically) persisted object. In the later case, using `dol`'s convention of interfacing with `typing.Mapping` (and assuming the `func_info` contains the `func_id`) we'd use a `json_schema_store` where we could do `json_schema_store[func_id]` or a `json_schema_store[func_id, schema_config]`.


<img width="386" alt="image" src="https://github.com/otosense/oui/assets/1906276/f5133403-5ca3-44d7-8ab4-7e9b217aded6">


### Adding conventions

Finally, if the configs contain a lot of parameters, it may be burdensome to have to fully express them anytime we'd want to do something else than the default. In that case, we can use the convention/configuration design, whereby:
* we allow the (dev) user to only override the parts of the underlying convention that they want to override, using the "conventional configuration" as a fall back
* we allow the convention to be changed as well

The diagram below makes this design explicit. 

<img width="603" alt="image" src="https://github.com/otosense/oui/assets/1906276/9c1b6e7f-bf17-44da-afd2-64de5431e730">

**Comment by valentin-feron** [comment]
Sathish's feedback and doubts:

"as far as i understood that
from end API ill receive the JSON / Metadata (func_info based on the flow diagram)
from that ill prepare the react component to show the data in table format using rjsf node module
once the user filled, using end point i need to pass the data to respective endpoint
and i need to display the output of the response in the UI again
this understood form the Steps which is mentioned in that.
my doubt is !
- is the API endpoints are ready?
- what kind of metadata/ JSON structure will be prepared and passed to the UI
- what kind of response (structure) that backend will support (payload structure)
(edited)

Also the Back-end support this part will be taken care by you right?

here 2 things
ui_configs and JSON_schema based on these the Form will bre created.
my doubts is!
- in that JSON Schema will be prepared by backend and pass to me
- but the ui_configs where ill get that? do i need to create (dev user)

from Thor's word i understood this much only Valentin!
please help me out if i missed anything!!"

**Reply by valentin-feron** [reply]
When we mention the API, we are in the context of the platform application. It is mentioned in the discussion to describe the use case when the user need to execute a function in the backend but from the front end. Now, regarding the func caller component itself, you understand that it does not depend on any API right? Meaning that he will never take an API url or any other api specification as an input. Instead, and it is what you've implemented so far, it takes a function as an input. Then this function can do anything (call an api, stay in the front end, etc.), but it is not the concern of the func caller component.

Regarding the json schema, we already made the choice to use the one provided by rjsf. But though, we'd like to have the flexibility for any dev user to define his own schema. To do so, we can add a transformer to the props so it transforms the input schema (user made) to the rjsf schema. So, again, dependency injection, the user provides the transformer function to be used to transfrom the schema object to comply with the initial format (the one from rjsf).

Regarding the output, it is pretty the same logic than above. We want to give the user a way to define how the output will be displayed. So we need another prop (another function) that will take the raw output (the one from the main function) and transform it accordingly (returning a JSX.Element that will be displayed). This transformer might not only depend on the output of the main function but also the schema and/or other props. In that case, we need the user to define his own transformer factory

More generally, this is what we call "ingress/egress". Take a look at the [`wrap`](https://github.com/i2mint/i2/blob/110c7050c4065d6f95ed0a3f5ab251664d25c430/i2/wrapper.py#L351) function/decorator. We might want to implement something similar in js for our components, to transform the props (ingress) and the `render` function's output (egress). Let's implement a specific solution for the function caller first, letting it open-close to a future general solution, using this same (or similar) ingress/egress mechanism.

**Comment by thorwhalen** [comment]
Some content from the original post...

## Technology search

### ReactJS frameworks for forms

These are a few frameworks we could look into:

* **Formik**: Formik is a widely used library that simplifies the process of building, validating, and managing forms in ReactJS. It provides components and hooks that make dealing with forms a breeze. Formik is great because it helps manage form state and form submission, and it provides support for validation and error messages out of the box.
* **React Hook Form**: This is a relatively new library that uses React Hooks for all its functionality. It aims to provide a performant, flexible, and less verbose way to manage forms in React applications. It supports form validation and is compatible with both JavaScript and TypeScript.
* **Redux Form**: If you're already using Redux in your application, Redux Form can be a good choice. It manages your form state in Redux, allowing you to keep your form state in sync with the rest of your application state.
* **react-jsonschema-form**: This is a powerful library that generates a form based on a JSON Schema. If you have a complex form and you want to generate it dynamically based on the server response, this can be a very useful tool.
* **Final Form**: Another library for managing form state in React. Like Formik and React Hook Form, it supports form validation and a wide variety of field types.

### react-jsonschema-form

The most promising one, for our purposes, seems to be `react-jsonschema-form`.

`react-jsonschema-form` (also known as rjsf) is a React component for building Web forms from JSON Schema. It has various features that make creating forms much simpler:
1.	Automatic Form Generation: It can automatically generate a React form interface from a JSON Schema, which describes the structure and validation requirements of the data your form should collect.
2.	Built-in Validation: JSON Schema offers declarative validation, and react-jsonschema-form provides automatic input validation and error messages. When a form is submitted, it checks that the collected data is valid against the provided schema.
3.	UI Customization: You can customize how your forms look by specifying a uiSchema object. This allows you to customize the order of fields, field labels, placeholder text, and many other properties of your form fields.
4.	Widgets and Field Templates: react-jsonschema-form allows you to define custom widgets and field templates. This means you can replace the standard HTML form controls (like text inputs, checkboxes, select dropdowns, etc.) with your own React components to fully control the rendering of your fields.
5.	Integration with Popular Libraries: react-jsonschema-form also supports integration with popular UI libraries like Bootstrap and Material-UI for the form layout, which can help to speed up the development process.
6.	Form Data and Schema Relationship: It allows you to use schema dependencies which means based on the value of one field, another field can be added to the form.
7.	Array and Object Support: It natively supports array and object types, and provides a straightforward interface for adding and removing items.
By using react-jsonschema-form, you can separate the form design from the UI, making your React code cleaner and simpler to maintain.

## Example (not complete) function to convert func to json-schema

```python
import inspect
from typing import Callable

def function_to_json_schema(func: Callable):
    # Fetch function metadata
    sig = inspect.signature(func)
    parameters = sig.parameters

    # Start building the JSON schema
    schema = {
        "title": func.__name__,
        "type": "object",
        "properties": {},
        "required": []
    }

    # Build the schema for each parameter
    for name, param in parameters.items():
        # Set the type based on the annotation
        if param.annotation is int:
            field = {"type": "integer"}
        elif param.annotation is float:
            field = {"type": "number"}
        elif param.annotation is bool:
            field = {"type": "boolean"}
        else:
            field = {"type": "string"}

        # If there's a default value, add it to the schema
        if param.default is not inspect.Parameter.empty:
            field["default"] = param.default
        else:
            schema["required"].append(name)

        # Add the field to the schema
        schema["properties"][name] = field

    return schema
```

## Example of code to inspire React component

```js
import React, { useState, useEffect } from 'react';
import Form from "@rjsf/core";
import axios from 'axios';

const MyForm = () => {
  const [selectedFunction, setSelectedFunction] = useState(null);
  const [functions, setFunctions] = useState([]);
  const [schema, setSchema] = useState(null);
  const [result, setResult] = useState(null);

  // Fetch the list of functions when the component mounts
  useEffect(() => {
    axios.get('/api/functions')
      .then(response => {
        setFunctions(response.data);
      });
  }, []);

  // Fetch the schema whenever the selected function changes
  useEffect(() => {
    if (selectedFunction) {
      axios.get(`/api/schemas/${selectedFunction}`)
        .then(response => {
          setSchema(response.data);
        });
    }
  }, [selectedFunction]);

  // Handle form submission
  const onSubmit = ({formData}) => {
    axios.post(`/api/call/${selectedFunction}`, formData)
      .then(response => {
        setResult(response.data);
      });
  };

  return (
    <div>
      <h1>Select a function</h1>
      <select value={selectedFunction} onChange={e => setSelectedFunction(e.target.value)}>
        {functions.map(func => (
          <option key={func} value={func}>{func}</option>
        ))}
      </select>
      
      {schema && (
        <div>
          <h1>Fill the form</h1>
          <Form schema={schema} onSubmit={onSubmit} />
        </div>
      )}

      {result && (
        <div>
          <h1>Result</h1>
          <pre>{JSON.stringify(result, null, 2)}</pre>
        </div>
      )}
    </div>
  );
};

export default MyForm;
```

**Comment by thorwhalen** [comment]
Have a single `egress` be added after the use input form, just before the submit button. 

The `egress` can be pointing to a JS function or a backend python function (which doesn't even need a http service). 

Note that the user can add more egresses, that will be chained. 

Each `egress` input field will be a rich (filterable) select box which is sourced by a `get_egresses` function that itself is sourced by some store (key-value object -- keys being displayed and values being the egress function associated to them).

In an app (example, oasis), the list (the store) of available functions there is determined by a special filter (that the user can change). 

In the image below, we show one egress with the default `raw_display`.

<img width="840" alt="image" src="https://github.com/otosense/oui/assets/1906276/e0572410-1cf0-43bf-a124-26f78873c5c8">

**Comment by thorwhalen** [comment]
### In react-json-schema-form, how can I source a select box from an API call?

In `react-jsonschema-form` (often abbreviated as RJSF), you can populate a `select` box with options sourced from an API call by using the `ui:options` field in the UI schema to specify a dynamic list of options. However, RJSF doesn't provide a built-in way to make API calls directly. You'll need to fetch the data yourself and then update the form's schema or UI schema accordingly.

Here's a basic example to demonstrate how you can achieve this:

### Step 1: Fetch Data from API

First, you'll need to fetch the data from the API. You can use `fetch`, `axios`, or any other method to get the data.

```javascript
import React, { useEffect, useState } from 'react';
import Form from 'react-jsonschema-form';

const App = () => {
  const [formData, setFormData] = useState({});
  const [selectOptions, setSelectOptions] = useState([]);

  useEffect(() => {
    // Fetch data from API
    fetch('https://api.example.com/items')
      .then((response) => response.json())
      .then((data) => {
        // Transform data to { label, value } format if needed
        const options = data.map((item) => ({
          label: item.name,
          value: item.id,
        }));
        setSelectOptions(options);
      });
  }, []);
  // ...
}
```

### Step 2: Update Schema or UI Schema

After fetching the data, update the `enum` and `enumNames` fields in the schema or the UI schema.

```javascript
// JSON Schema
const schema = {
  type: 'object',
  properties: {
    item: {
      type: 'string',
      enum: selectOptions.map((option) => option.value),
      enumNames: selectOptions.map((option) => option.label),
    },
  },
};

// UI Schema
const uiSchema = {
  item: {
    'ui:widget': 'select',
  },
};
```

### Step 3: Render the Form

Finally, render the form using the updated schema and UI schema.

```javascript
return (
  <Form
    schema={schema}
    uiSchema={uiSchema}
    formData={formData}
    onChange={(e) => setFormData(e.formData)}
  />
);
```

Putting it all together:

```javascript
import React, { useEffect, useState } from 'react';
import Form from 'react-jsonschema-form';

const App = () => {
  const [formData, setFormData] = useState({});
  const [selectOptions, setSelectOptions] = useState([]);

  useEffect(() => {
    fetch('https://api.example.com/items')
      .then((response) => response.json())
      .then((data) => {
        const options = data.map((item) => ({
          label: item.name,
          value: item.id,
        }));
        setSelectOptions(options);
      });
  }, []);

  const schema = {
    type: 'object',
    properties: {
      item: {
        type: 'string',
        enum: selectOptions.map((option) => option.value),
        enumNames: selectOptions.map((option) => option.label),
      },
    },
  };

  const uiSchema = {
    item: {
      'ui:widget': 'select',
    },
  };

  return (
    <Form
      schema={schema}
      uiSchema={uiSchema}
      formData={formData}
      onChange={(e) => setFormData(e.formData)}
    />
  );
};

export default App;
```

This example demonstrates how to populate a `select` box in `react-jsonschema-form` using data fetched from an API.

**Comment by thorwhalen** [comment]
### Can I map some fields to custom input JS elements in react-jsonschema-form?

Yes, you can map fields to custom input elements in `react-jsonschema-form` (RJSF). The library provides a way to use custom widgets and fields via the `fields` and `widgets` props. Here's how you can do it:

### Custom Widgets

You can create custom widgets and pass them to the `Form` component via the `widgets` prop. A widget is essentially a custom React component that receives certain props from the RJSF library.

Here's an example of a custom text input widget:

```javascript
const CustomTextInput = (props) => {
  return (
    <input
      type="text"
      value={props.value}
      required={props.required}
      onChange={(event) => props.onChange(event.target.value)}
    />
  );
};
```

To use this custom widget, you would pass it in the `widgets` prop like so:

```javascript
import Form from 'react-jsonschema-form';

const widgets = {
  customTextInput: CustomTextInput,
};

<Form schema={schema} uiSchema={uiSchema} widgets={widgets} />
```

And in your UI schema, you would specify that you want to use this custom widget:

```javascript
const uiSchema = {
  textField: {
    "ui:widget": "customTextInput",
  },
};
```

### Custom Fields

Similarly, you can create custom fields. A custom field is a bit more involved than a custom widget because it allows you to control the entire layout of a field, including its label, the input element, and any validation messages.

Here's an example of a custom field:

```javascript
const CustomField = (props) => {
  return (
    <div>
      <label>{props.label}</label>
      <input
        type="text"
        value={props.formData}
        onChange={(event) => props.onChange(event.target.value)}
      />
      {props.rawErrors && (
        <ul>
          {props.rawErrors.map((error, i) => (
            <li key={i}>{error}</li>
          ))}
        </ul>
      )}
    </div>
  );
};
```

To use this custom field, you would pass it in the `fields` prop like so:

```javascript
import Form from 'react-jsonschema-form';

const fields = {
  customField: CustomField,
};

<Form schema={schema} uiSchema={uiSchema} fields={fields} />
```

And in your UI schema, you would specify that you want to use this custom field:

```javascript
const uiSchema = {
  textField: {
    "ui:field": "customField",
  },
};
```

By using custom widgets and fields, you can have complete control over the rendering of inputs in `react-jsonschema-form`, allowing you to use any custom input elements you'd like.
